### ISSUE TITLE
ERROR: AADSTS700024: Client assertion is not within its valid time range

### ISSUE THREAD

**Issue by guidoiaquinti** [Azure/login#180]
👋 Hi! I’m using the `azure/login` action for CD and I’m often getting an error when the job execution takes more than a couple of minutes from the initial login. 

```
ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2021-11-26T09:44:41.0193003Z, expiry time of assertion 2021-11-26T09:31:56.0000000Z.
```

Is there a way to extend the validity of the token via a parameter? The `aws-actions/configure-aws-credentials` action does it via a  `role-duration-seconds` flag:

> The default session duration is 6 hours when using an IAM User to assume an IAM Role [..] . If you would like to adjust this you can pass a duration to role-duration-seconds, but the duration cannot exceed the maximum that was defined when the IAM Role was created.

>

> ref: https://github.com/aws-actions/configure-aws-credentials/blob/master/README.md#assuming-a-role

Thank you! 🙇

**Comment by BALAGA-GAYATRI** [comment]
@kjyam98 @N-Usha @udayxhegde Is this Doable in Azure?

**Comment by udayxhegde** [comment]
@BALAGA-GAYATRI I am not sure I understand the question: what is Doable in Azure? The issue seems to be that the github token has expired when it is presented to Azure AD

**Comment by guidoiaquinti** [comment]
@udayxhegde the problem I’m trying to solve is customise the expiration time of the token. This is useful if you plan to run operations on the Azure platform with timeframes longer than the standard validity.

Example:

1. Provisioning of the AKS cluster
2. deploy a stack 
3. execute e2e tests
4. destruction of the AKS cluster

If the timeframe between steps 1 and 4 is longer than the maximum validity of the token, you will encounter an error similar to the one I posted above.

It would be useful to customise the token validity via `azure/login`. This is something that other cloud platforms already support.

Example:
```
    - name: Configure AWS Credentials
      uses: aws-actions/configure-aws-credentials@v1
      with:
        role-to-assume: arn:aws:iam::123456789100:role/my-github-actions-role
        role-duration-seconds: 3600
```

**Comment by BALAGA-GAYATRI** [comment]
Hey @guidoiaquinti, This is currently not being implemented in OIDC login mechanism. If you want to leverage the auto refresh mechanism of token, you may use the SPN login by setting your deployment credentials using - https://github.com/Azure/login#configure-a-service-principal-with-a-secret
In this way you will not face token expiration issue.

**Comment by guidoiaquinti** [comment]
> Hey @guidoiaquinti, This is currently not being implemented in OIDC login mechanism. If you want to leverage the auto refresh mechanism of token, you may use the SPN login by setting your deployment credentials using - https://github.com/Azure/login#configure-a-service-principal-with-a-secret In this way you will not face token expiration issue.

Thank you for your reply Balaga 🙇 

Any plan to add this feature to the Azure roadmap? I'm trying to avoid generating and hardcoding long-term secrets if possible.

**Comment by BALAGA-GAYATRI** [comment]
Yes @guidoiaquinti. The refresh token mechanism is yet to be implemented.

**Comment by guidoiaquinti** [comment]
I understand but is there any plan to implement it?

**Comment by BALAGA-GAYATRI** [comment]
Yes. But we can't commit to a specific timeline for this. We will keep you informed.

**Comment by kjyam98** [comment]
@BALAGA-GAYATRI please assign to udayxhegde

**Comment by udayxhegde** [comment]
@kjyam98, that is not the right assignment

> @BALAGA-GAYATRI please assign to udayxhegde

@kjyam98 , @BALAGA-GAYATRI: what are you expecting by assigning to me? did you have a specific design in mind that you need my help?

**Comment by BALAGA-GAYATRI** [comment]
> @kjyam98, that is not the right assignment
> 
> > @BALAGA-GAYATRI please assign to udayxhegde
> 
> @kjyam98 , @BALAGA-GAYATRI: what are you expecting by assigning to me? did you have a specific design in mind that you need my help?

@udayxhegde 
Sorry for the confusion. This was already closed over email. We will be working with the CLI team for the changes required in CLI and in Action for the auto-refresh mechanism for addressing the same. Have assigned back the issue to my team.

**Comment by zoebosschaart** [comment]
Following this issue as I experience the same.

**Comment by vslatten** [comment]
Also experiencing this issue for a long-running workflow.

**Comment by christianacca** [comment]
Solution still required

**Comment by christianacca** [comment]
Required please

**Comment by christianacca** [comment]
Royal pain in the butt this one. Requires splitting up any long running ARM / scripts to login between each split out piece

**Comment by matt-FFFFFF** [comment]
Current OIDC token lifetime seems to be about 4 minutes, this is WAY too short and should have a sensible default. Consider that GitHub Actions jobs can run for [6 hours](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits).

@BALAGA-GAYATRI Is there an update here? Do you have a link to the upstream issue on Azure CLI?

Thanks

**Comment by christianacca** [comment]
Still required

**Comment by timkowal** [comment]
Bump.

Using the alternate login method is less secure than OIDC.

**Comment by sandrochristiaan** [comment]
Bump.

Any update on when the solution might be implemented?

Please notify, so we can determine to go with the subpar sp secret approach...

Thanks.

**Comment by christianacca** [comment]
bumo

**Comment by N-Usha** [comment]
Thanks for your patience on this Folks.. Can u please confirm if you have explored the Azure settings to modify the life time of the access token: https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens#access-token-lifetime.

While we acknowledge that this is not a permanent solution, we believe that it can help if the workflow running time is slightly more than the default access tokens' lifetime.  Thanks!

**Comment by sandrochristiaan** [comment]
Hello @N-Usha I have modified the token lifetime using the Microsoft documentation mentioned. 
It does not make any difference at all, deployments will fail due to the token expiring after a short period of time. 

I opened a case with Microsoft on this. They have told me that the token lifetime policy will only apply to the SP token and not the assertion from the federated credentials.
An auto-refresh mechanism is what is needed to fix this issue.

I do hope this to be fixed soon, since this really is a way better solution than instead of using a service principal with secrets, as described in GitHub documentation: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#benefits-of-using-oidc

Can you provide a timeline on when to expect a fix?
Or even an answer if it will get fixed?

Thanks,
Sandro

**Comment by N-Usha** [comment]
Thanks @sandrochristiaan. We have it on our backlog to address the refresh issue and will have it addressed soon. Will surely keep you posted on the timelines.

Yes, we definitely recommend OIDC for authentication as it helps avoid transacting with cloud credentials as secrets. And we are committed to making the experience better for our customers. Thanks for your support again.

**Comment by sandrochristiaan** [comment]
bump

**Comment by matt-FFFFFF** [comment]
Still necessary

**Comment by matt-FFFFFF** [comment]
Still necessary (this is fun 👍)

**Comment by RenatoMartins-tomtom** [comment]
bump again. More than one year in the making? Plenty of docs about github actions + azure using OIDC (perfect, no need to rotate credentials). And the pipeline can run for negligible < 10 minutes? 
https://learn.microsoft.com/en-us/azure/developer/github/connect-from-azure?tabs=azure-portal%2Cwindows

https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-azure

Token policies, as mentioned, have no effect. No defaults changed on the tenant, and the token is valid for 1 hour... https://learn.microsoft.com/en-us/azure/active-directory/develop/active-directory-configurable-token-lifetimes#access-id-and-saml2-token-lifetime-policy-properties

**Comment by sandrochristiaan** [comment]
@RenatoMartins-tomtom I totally agree, it is a bit weird the difference between reality and what is being propagated in the documentation.

I would be quite happy to just have a 'draft' timeline on when to expect possible work to be done on this issue :| 

Please can someone please provide a real answer on what we can expect?

**Comment by udayxhegde** [comment]
The tokens you get for the Application registration should be valid for an hour, not less than that. If anyone has tokens expiring less than an hour, can you provide more details on your action and where it is expiring?

Also: federation on managed identities is in public preview, and these tokens should work even if the actions take 5 hours. Please see if that addresses your scenario in a better way

**Comment by RenatoMartins-tomtom** [comment]
Hi @udayxhegde . The error I get is similar to this:
```
ClientAssertionCredential authentication failed: A
     | configuration issue is preventing authentication - check the
     | error message from the server for details. You can modify the
     | configuration in the application registration portal. See
     | https://aka.ms/msal-net-invalid-client for details.  Original
     | exception: AADSTS700024: Client assertion is not within its
     | valid time range. Current time: 2022-10-26T06:11:15.1049397Z,
     | assertion valid from 2022-10-26T06:01:13.0000000Z, expiry time
     | of assertion 2022-10-26T06:06:13.0000000Z. Review the
     | documentation at
```

The action is as simple as this
```
 runs-on: ubuntu-latest
    steps:

    - name: Checkout self
      uses: actions/checkout@v2

    - name: Login to Azure 
      uses: azure/login@v1
      with:
        client-id: ${{ secrets.AZURE_CLIENT_ID }}
        tenant-id: ${{ secrets.AZURE_TENANT_ID }}
        enable-AzPSSession: true 
        allow-no-subscriptions: true
    
    - name: Scan eligible groups for member count
      shell: pwsh
      env:
        TENANT_ID: xyz
        STORAGE_ACCOUNT_NAME: 'pimcompliancedata'
        STORAGE_CONTAINER_NAME: 'pim-daily-scan-results'
        
      run: |
        .\scripts\scan-member-count.ps1 -storageAccountName $env:STORAGE_ACCOUNT_NAME -storageAccountContainer $env:STORAGE_CONTAINER_NAME

```
And the PS script being run is just calling a bunch of Get-AzAdGroupMember for all AD groups starting with a given prefix (in a foreach-object -parallel). And this happens every time the action takes more than 10 mins. Tenant settings are per default (which should make the token valid for 1 hour). OIDC federation is setup as the documentation linked above asks to do ("scenario" is "github actions deploying azure resources").

**Comment by arnemorten** [comment]
I experienced this problem when using az acr login. We build multiple images so sometimes it takes more than 5 minutes before we can push the next image. We solved it by just running az acr login multiple times. Did not see this behavior before we switched to federated credentials with azure/login@v1. 

 @RenatoMartins-tomtom I think if you can workaround this issue by using the az cli commands instead of Get-AzAdGroupMember. I believe az cli automatically refreshes the tokens for you.

**Comment by cwe1ss** [comment]
Here's my error with a sample workflow. As you can see, this does just log in and wait for 11 minutes between two calls to the Az PowerShell module.

```yaml
name: 'TEST'

on:
  workflow_dispatch:

permissions:
  id-token: write
  contents: read

jobs:

  deploy:
    runs-on: ubuntu-latest
    environment: platform

    steps:

    - uses: actions/checkout@v3

    - uses: azure/login@v1
      with:
          client-id: ${{ secrets.AZURE_CLIENT_ID }}
          tenant-id: ${{ secrets.AZURE_TENANT_ID }}
          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
          enable-AzPSSession: true

    - name: 'Deploy Azure resources'
      uses: azure/powershell@v1
      with:
        inlineScript: |
          "Calling Get-AzVm"
          Get-AzVm
          "Sleep start"
          Start-Sleep -Seconds 660
          "Sleep stop"
          "Calling Get-AzVm"
          Get-AzVm
          "Done"
        azPSVersion: "8.2.0"
```

Output from azure/login:
<details><summary>Show output</summary>

2022-10-28T08:12:27.5734334Z ##[group]Run azure/login@v1
2022-10-28T08:12:27.5734597Z with:
2022-10-28T08:12:27.5734991Z   client-id: ***
2022-10-28T08:12:27.5735275Z   tenant-id: ***
2022-10-28T08:12:27.5735565Z   subscription-id: ***
2022-10-28T08:12:27.5735817Z   enable-AzPSSession: true
2022-10-28T08:12:27.5736072Z   environment: azurecloud
2022-10-28T08:12:27.5736320Z   allow-no-subscriptions: false
2022-10-28T08:12:27.5736614Z   audience: api://AzureADTokenExchange
2022-10-28T08:12:27.5736859Z ##[endgroup]
2022-10-28T08:12:28.6273343Z Using OIDC authentication...
2022-10-28T08:12:28.6974981Z Federated token details: 
2022-10-28T08:12:28.6975674Z  issuer - https://token.actions.githubusercontent.com 
2022-10-28T08:12:28.6976257Z  subject claim - repo:cwe1ss/msa-template:environment:platform
2022-10-28T08:12:28.6981058Z [command]/usr/bin/az cloud set -n azurecloud
2022-10-28T08:12:30.1968287Z Done setting cloud: "azurecloud"
2022-10-28T08:12:31.7047442Z Running Azure PS Login
2022-10-28T08:12:31.7089422Z [command]/usr/bin/pwsh -Command try {
2022-10-28T08:12:31.7090020Z             $ErrorActionPreference = "Stop"
2022-10-28T08:12:31.7090580Z             $WarningPreference = "SilentlyContinue"
2022-10-28T08:12:31.7092921Z             $output = @{}
2022-10-28T08:12:31.7094749Z             $data = Get-Module -Name Az.Accounts -ListAvailable | Sort-Object Version -Descending | Select-Object -First 1
2022-10-28T08:12:31.7095423Z             $output['AzVersion'] = $data.Version.ToString()
2022-10-28T08:12:31.7096004Z             $output['Success'] = "true"
2022-10-28T08:12:31.7096670Z         }
2022-10-28T08:12:31.7096935Z         catch {
2022-10-28T08:12:31.7097400Z             $output['Error'] = $_.exception.Message
2022-10-28T08:12:31.7097661Z         }
2022-10-28T08:12:31.7097962Z         return ConvertTo-Json $output
2022-10-28T08:13:06.3608603Z {
2022-10-28T08:13:06.3608940Z   "AzVersion": "2.9.1",
2022-10-28T08:13:06.3609586Z   "Success": "true"
2022-10-28T08:13:06.3609828Z }
2022-10-28T08:13:06.4372988Z [command]/usr/bin/pwsh -Command try {
2022-10-28T08:13:06.4373309Z             $ErrorActionPreference = "Stop"
2022-10-28T08:13:06.4373776Z             $WarningPreference = "SilentlyContinue"
2022-10-28T08:13:06.4374132Z             $output = @{}
2022-10-28T08:13:06.4374559Z             Clear-AzContext -Scope Process;
2022-10-28T08:13:06.4422814Z              Clear-AzContext -Scope CurrentUser -Force -ErrorAction SilentlyContinue;Connect-AzAccount -ServicePrincipal -ApplicationId '***' -Tenant '***' -FederatedToken '***'                      -Environment 'azurecloud' | out-null;Set-AzContext -SubscriptionId '***' -TenantId '***' | out-null;
2022-10-28T08:13:06.4423597Z             $output['Success'] = "true"
2022-10-28T08:13:06.4423839Z         }
2022-10-28T08:13:06.4510346Z         catch {
2022-10-28T08:13:06.4510924Z             $output['Error'] = $_.exception.Message
2022-10-28T08:13:06.4511192Z         }
2022-10-28T08:13:06.4511507Z         return ConvertTo-Json $output
2022-10-28T08:13:12.5383688Z {
2022-10-28T08:13:12.5384284Z   "Success": "true"
2022-10-28T08:13:12.5384636Z }
2022-10-28T08:13:12.6247805Z Azure PowerShell session successfully initialized
2022-10-28T08:13:12.6248618Z Login successful.
</details>

Output from azure/powershell:

<details><summary>Show output</summary>
2022-10-28T08:13:12.6351560Z ##[group]Run azure/powershell@v1
2022-10-28T08:13:12.6351822Z with:
2022-10-28T08:13:12.6352208Z   inlineScript: "Calling Get-AzVm"
Get-AzVm
"Sleep start"
Start-Sleep -Seconds 660
"Sleep stop"
"Calling Get-AzVm"
Get-AzVm
"Done"

2022-10-28T08:13:12.6352593Z   azPSVersion: 8.2.0
2022-10-28T08:13:12.6352854Z   errorActionPreference: Stop
2022-10-28T08:13:12.6353125Z   failOnStandardError: false
2022-10-28T08:13:12.6353568Z   githubToken: ***
2022-10-28T08:13:12.6353792Z env:
2022-10-28T08:13:12.6354008Z   AZURE_HTTP_USER_AGENT: 
2022-10-28T08:13:12.6354261Z   AZUREPS_HOST_ENVIRONMENT: 
2022-10-28T08:13:12.6354500Z ##[endgroup]
2022-10-28T08:13:12.8442153Z Validating inputs
2022-10-28T08:13:12.8710862Z [command]/usr/bin/pwsh -NoLogo -NoProfile -NonInteractive -Command Test-Path (Join-Path /usr/share az_*)
2022-10-28T08:13:13.2067027Z True
2022-10-28T08:13:13.8151481Z [command]/usr/bin/pwsh -NoLogo -NoProfile -NonInteractive -Command 
2022-10-28T08:13:13.8152170Z             $prevProgressPref = $ProgressPreference
2022-10-28T08:13:13.8153228Z             $ProgressPreference = 'SilentlyContinue'
2022-10-28T08:13:13.8153764Z             Expand-Archive -Path /usr/share/az_8.2.0.zip -DestinationPath /usr/share
2022-10-28T08:13:13.8154157Z             $ProgressPreference = $prevProgressPref
2022-10-28T08:13:17.2987612Z Module Az 8.2.0 installed from hostedAgentGHRelease
2022-10-28T08:13:17.2989446Z Initializing Az Module
2022-10-28T08:13:17.3004255Z Initializing Az Module Complete
2022-10-28T08:13:17.3006143Z Running Az PowerShell Script
2022-10-28T08:13:17.3021054Z [command]/usr/bin/pwsh -NoLogo -NoProfile -NonInteractive -Command /home/runner/work/_temp/ec0866d1-cdc6-473a-a3a8-a984c5e898ca.ps1
2022-10-28T08:13:17.5657741Z Calling Get-AzVm
2022-10-28T08:13:21.7046544Z Sleep start
2022-10-28T08:24:21.7076491Z Sleep stop
2022-10-28T08:24:21.7077432Z Calling Get-AzVm
2022-10-28T08:24:22.1724491Z [31;1mGet-AzVM: [0m/home/runner/work/_temp/ec0866d1-cdc6-473a-a3a8-a984c5e898ca.ps1:8
2022-10-28T08:24:22.1724938Z [36;1mLine |
2022-10-28T08:24:22.1725247Z [36;1m   8 | [0m [36;1mGet-AzVm[0m
2022-10-28T08:24:22.1725542Z [36;1m     | [31;1m ~~~~~~~~
2022-10-28T08:24:22.1725993Z [31;1m[36;1m     | [31;1mYour Azure credentials have not been set up or have expired, please run
2022-10-28T08:24:22.1726519Z [36;1m     | [31;1mConnect-AzAccount to set up your Azure credentials. A configuration
2022-10-28T08:24:22.1727036Z [36;1m     | [31;1missue is preventing authentication - check the error message from the
2022-10-28T08:24:22.1727552Z [36;1m     | [31;1mserver for details. You can modify the configuration in the application
2022-10-28T08:24:22.1728095Z [36;1m     | [31;1mregistration portal. See https://aka.ms/msal-net-invalid-client for
2022-10-28T08:24:22.1728901Z [36;1m     | [31;1mdetails.  Original exception: AADSTS700024: Client assertion is not
2022-10-28T08:24:22.1729401Z [36;1m     | [31;1mwithin its valid time range. Current time: 2022-10-28T08:24:21.8673176Z,
2022-10-28T08:24:22.1729874Z [36;1m     | [31;1massertion valid from 2022-10-28T08:12:29.0000000Z, expiry time of
2022-10-28T08:24:22.1730344Z [36;1m     | [31;1massertion 2022-10-28T08:17:29.0000000Z. Review the documentation at
2022-10-28T08:24:22.1731256Z [36;1m     | [31;1mhttps://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials . Trace ID: 97fd40c6-56db-4450-81a4-5124da090800 Correlation ID: bf08574a-c5b8-414c-a061-10f576bcb360 Timestamp: 2022-10-28 08:24:21Z
2022-10-28T08:24:22.1731820Z [0m
2022-10-28T08:24:22.2577355Z ##[error]Error: The process '/usr/bin/pwsh' failed with exit code 1
</details>

The error in a more readable form:
```
Get-AzVM: /home/runner/work/_temp/ec0866d1-cdc6-473a-a3a8-a984c5e898ca.ps1:8
Line |
   8 |  Get-AzVm
     |  ~~~~~~~~
     | Your Azure credentials have not been set up or have expired, please run
     | Connect-AzAccount to set up your Azure credentials. A configuration
     | issue is preventing authentication - check the error message from the
     | server for details. You can modify the configuration in the application
     | registration portal. See https://aka.ms/msal-net-invalid-client for
     | details.  Original exception: AADSTS700024: Client assertion is not
     | within its valid time range. Current time: 2022-10-28T08:24:21.8673176Z,
     | assertion valid from 2022-10-28T08:12:29.0000000Z, expiry time of
     | assertion 2022-10-28T08:17:29.0000000Z. Review the documentation at
     | https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials . Trace ID: 97fd[40](https://github.com/cwe1ss/msa-template/actions/runs/3343879514/jobs/5537575948#step:4:41)c6-56db-4450-81a4-5124da090800 Correlation ID: bf08574a-c5b8-[41](https://github.com/cwe1ss/msa-template/actions/runs/3343879514/jobs/5537575948#step:4:42)4c-a061-10f576bcb360 Timestamp: 2022-10-28 08:24:21Z
```

**Comment by matt-FFFFFF** [comment]
Another issue here with the token validity here, the token only being valid for 5 mins!

Using Azure/Login with Service Principal:

```
    subscription.go:24: cancelling subscription 1ad7747d-5335-4322-a565-392903a10534
    subscriptionDeploy_test.go:96: cannot cancel subscription: subscription 1ad7747d-5335-4322-a565-392903a10534 does not exist or cannot successfully check, cannot get subscription, DefaultAzureCredential: failed to acquire a token.
        Attempted credentials:
        	EnvironmentCredential: missing environment variable AZURE_CLIENT_ID
        	ManagedIdentityCredential: IMDS token request timed out
        	AzureCLICredential: ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2022-10-31T12:44:19.4908573Z, assertion valid from 2022-10-31T11:49:11.0000000Z, expiry time of assertion 2022-10-31T11:54:11.0000000Z. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials .
```

**Comment by rwaal** [comment]
We are having the exact same issue as @cwe1ss mentioned [here](https://github.com/Azure/login/issues/180#issuecomment-1294707313). 
@BALAGA-GAYATRI, are you still involved with this issue? Any ideas on timelines? Thanks in advance.

**Comment by BALAGA-GAYATRI** [comment]
There are a few things that we came across while looking at this issue.

1. The known limitation of access-token expiry and absence of a refresh mechanism. This is one potential cause for the failure of long-running workflows which go beyond the default validity of the access-token. For the default values please check this - https://learn.microsoft.com/en-us/azure/active-directory/develop/access-tokens#access-token-lifetime
2. There are a few comments on workflows failing after 5 or 10 mins for the powershell case. We have noticed that 
azure PowerShell fetches an access token from MSAL when each cmdlet execution. For client assertion (Github token), azure PowerShell stores assertion and fetches access token using the github token when each cmdlet is executed. This is causing the workflows to fail soon after the github token expires. This behavior only applies to the powershell case. In the Az CLI login, we don't observe this. We are following up with the respective teams to understand the problem better as this is not related to login action.

We shall post the status here for any new updates.

**Comment by BALAGA-GAYATRI** [comment]
A tracking issue for this has been opened in PowerShell for this. Link to the issue - https://github.com/Azure/azure-powershell/issues/20013

**Comment by RenatoMartins-tomtom** [comment]
Just today, noticed it will fail when using CLI as well. Well, the action calls a powershell script, but in there, all the commands are CLI commands (powershell commands are not AZ-Powershell ones, only foreach, write-out, or similar ones). After going for 13 mins in a loop, it would try an
```
az storage blob upload --file $fileName -c $storageAccountContainer --account-name $storageAccountName -n $fileName --auth-mode login
```

failing with the same error

```
ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2022-11-11T13:57:03.8472818Z, assertion valid from 2022-11-11T13:43:07.0000000Z, expiry time of assertion 2022-11-11T13:48:07.0000000Z. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials .
Trace ID: c8ca033e-5187-40aa-8b7c-dc8c7762e300
Correlation ID: ed3e4dfc-4dbd-4096-9dfe-c855164bb964
Timestamp: 2022-11-11 13:57:03Z
Interactive authentication is needed. Please run:
az login

```

and there's certainly no way to use az login...

**Comment by BALAGA-GAYATRI** [comment]
@RenatoMartins-tomtom Please share your workflow masking sensitive information. This can help us get better idea on the issue you are facing.

**Comment by leahy268** [comment]
Hey guys, 

This is a serious problem.  At the moment we are able to workaround the issue by splitting longer running PowerShell scripts up into smaller parts and logging in again.  

However that isn't always possible.

Please see my better software suggestion here:

https://bettersoftwaresuggestions.com/github/github-azure-login-allow-custom-expiry-time-for-oidc-token/

**Comment by RenatoMartins-tomtom** [comment]
Sure,

here is the workflow

```
name: scan-assessments
on:
  schedule:
    - cron: '0 6 * * *' # every day at 6:00
  workflow_dispatch:
permissions:
      id-token: write
      contents: read
jobs:
  pull-list-of-failed-assessments:
    runs-on: ubuntu-latest
    steps:

    - name: Checkout self
      uses: actions/checkout@v2

    - name: Login to Azure 
      uses: azure/login@v1
      with:
        client-id: ${{ secrets.AZURE_CLIENT_ID }}
        tenant-id: ${{ secrets.AZURE_TENANT_ID }}
        allow-no-subscriptions: true
    
    - name: Scan subscriptions for compliance
      shell: pwsh
      env:
        TENANT_ID: xyz
        STORAGE_ACCOUNT_NAME: 'pimcompliancedata'
        STORAGE_CONTAINER_NAME: 'defender-for-cloud-daily-scan-results'
        
      run: |
        .\scripts\scan-defender-assessments.ps1 -storageAccountName $env:STORAGE_ACCOUNT_NAME -storageAccountContainer $env:STORAGE_CONTAINER_NAME
````

The script it calls is below, Note it is purely cli based (powershell is only used as a scripting language, no Az-ps commands are ever called). Eventually fails after 10+ mins. The ones that use any az-ps commands will fail for sure:
```
[CmdletBinding()]
param(
    [Parameter(Mandatory = $true)]
    [string]$storageAccountName,

    [Parameter(Mandatory = $true)]
    [string]$storageAccountContainer

)

$date = (get-date).ToString("yyyy-MM-dd")
$fileName = "$date-defender-for-cloud-assessment-scan.csv"

# token valid until Jan 1st, 2023, allowing read, create and write for blobs.
$sasToken = "token"

az config set extension.use_dynamic_install=yes_without_prompt

# hard-coded values, as the standard has a single control, called 1
$stdName = "std"
$controlName = "1"

$subscriptions = az account list | ConvertFrom-Json
$output = @()
$output += "stage, bu, pu, team, subscriptionName, subscriptionId, passedAssessmentCount, failedassessmentCount, failedAssessmentDescription, failedAssessmentPassedResources, failedAssessmentSkippedResources, failedAssessmentFailedResources,"
$i = 0 
$total = $subscriptions.count
foreach ($subs in $subscriptions) {
	$i++
	"$i of $total"
	az account set --subscription $subs.id
	$cmdOutput =  az security regulatory-compliance-controls list --standard-name $stdName
	if ($cmdOutput -notlike "ERROR: *") {
		$mgHierarchy = (az graph query -q "ResourceContainers | where type =~ 'microsoft.resources/subscriptions' | extend mgParent = properties.managementGroupAncestorsChain | mv-expand with_itemindex=MGHierarchy mgParent | project subscriptionId, name, mgParent, MGHierarchy, mgParent.name" --subscriptions $subs.id | ConvertFrom-Json).data
		$mgName = $mgHierarchy.mgParent_name
		$prodNonProd = $mgName[$mgName.count - 3]
		$bu =  $mgName[$mgName.count - 4]
		$pu =  $mgName[$mgName.count - 5]
		$team = $mgName[$mgName.count - 6]
	

		$compliance = $cmdOutput | convertfrom-Json
		$line = """$prodNonProd"", ""$bu"", ""$pu"", ""$team"", ""$($subs.name)"", $($subs.id), $($compliance.passedassessments), $($compliance.failedassessments)"
		$assessments = az security regulatory-compliance-assessments list --control-name $controlName --standard-name $stdName
		$assessments = $assessments | convertfrom-json	
		$failedAssessments = $assessments | where {$_.state -eq "Failed"}
		foreach ($failedAssessment in $failedAssessments) {
			$line += "$($subs.name), $($subs.id), $($compliance.passedassessments), $($compliance.failedassessments),"
			$line += """$($failedAssessment.Description)"", $($failedAssessment.passedResources), $($failedAssessment.skippedResources), $($failedAssessment.failedResources)"
			$output += $line
		}
		
	}
}
$output | Out-File -FilePath $fileName

az storage blob upload --file $fileName -c $storageAccountContainer --account-name $storageAccountName -n $fileName --sas-token $sasToken 
```

**Comment by N-Usha** [comment]
Looping in @dcaro and @chasewilson from Azure PowerShell team, who are exploring a resolution for this issue in Az PS.

**Comment by dcaro** [comment]
@N-Usha this has been addressed by Azure CLI and will addressed in Azure PowerShell v9.2 to be released on 12/6/2022.

**Comment by krowlandson** [comment]
@BALAGA-GAYATRI are you able to please confirm if/when the fixes mentioned above will be available in this action? Thank you

**Comment by krowlandson** [comment]
Keeping this from being idle 👍🏻

**Comment by Jtango18** [comment]
2+ Years, surely someone has implemented a fix for this? When it takes 1hour+ to deploy a Dev Sku'd APIM instance, this makes OIDC a complete not starter.

**Comment by sonnysideup** [comment]
@Jtango18 You are correct. Except in limited cases, that typically exclude provisioning/decommissioning substantial infrastruction, this OIDC integration is unusable. How is this competitive with AWS/GCP? Not sure. 🤷🏼‍♂️

**Comment by N-Usha** [comment]
Thanks for your patience folks. Based on [this](https://github.com/Azure/login/issues/180#issuecomment-1336882031), Azure PowerShell has released their fix and this should have bumped up the default validity of the OIDC token issued for Service principals to 1hr.

Just to call out, Azure also supports OIDC with Managed identities and the default token validation time for this flow is 24 hrs - which should suffice most of the deployment use cases. 
Please refer to Microsoft's documentation at "[Workload identity federation](https://learn.microsoft.com/en-us/azure/active-directory/develop/workload-identity-federation)” which has more details about the Azure Workload Identity Federation (OIDC) support for Managed Identities.

Kindly request you to explore this and confirm if that experience works as expected for your scenarios. Thanks.

**Comment by sandrochristiaan** [comment]
> Thanks for your patience folks. Based on [this](https://github.com/Azure/login/issues/180#issuecomment-1336882031), Azure PowerShell has released their fix and this should have bumped up the default validity of the OIDC token issued for Service principals to 1hr.
> 
> Just to call out, Azure also supports OIDC with Managed identities and the default token validation time for this flow is 24 hrs - which should suffice most of the deployment use cases. Please refer to Microsoft's documentation at "[Workload identity federation](https://learn.microsoft.com/en-us/azure/active-directory/develop/workload-identity-federation)” and [Configure a user-assigned managed identity to trust an external identity provider (preview)](https://learn.microsoft.com/en-us/azure/active-directory/develop/workload-identity-federation-create-trust-user-assigned-managed-identity?pivots=identity-wif-mi-methods-azp) which has more details about the Azure Workload Identity Federation (OIDC) support for Managed Identities.
> 
> Kindly request you to explore this and confirm if that experience works as expected for your scenarios. Thanks.

Hello @N-Usha 

I just tested this and the lifetime is still not very long, as you can in the screenshot.

![image](https://user-images.githubusercontent.com/22005405/211870117-3abe2175-29f2-49b5-a842-37988aa72bb2.png)

Are there any specific configurations to be done?
I am using version 1.4.6 of the azure/login action

![image](https://user-images.githubusercontent.com/22005405/211870778-bed64d84-5807-4683-b808-91e60b222d53.png)

But it states AzVersion is 2.10.4, is this the correct version?

Please let me know if I need to configure something differently on my yaml files.

Cheers,
Sandro

**Comment by BALAGA-GAYATRI** [comment]
@dingmeng-xue @dcaro Please help with the above comment by @sandrochristiaan where the user is still facing the token expiry issue (5 mins instead of 1hr)

**Comment by dingmeng-xue** [comment]
Is there any scenario I don't know? I tried below actions. There is no problem. After login, it runs 20 minutes. The only difference is Az.Accounts version is 2.11.0. According to changelog of Az.Accounts, it doesn't matter.

```
        - name: OIDC Login to Azure Public Cloud with AzPowershell (enableAzPSSession true)
          uses: Azure/login@v1.4.6
          with:
            client-id: '***'
            tenant-id: '***'
            subscription-id: '***'
            enable-AzPSSession: true

        - name: 'Get resource group with PowerShell action'
          uses: azure/powershell@v1
          with:
             inlineScript: |
               Get-AzAccessToken
               for(($i =0);$i -lt 20; $i++) { $list = Get-AzResourceGroup; Write-Host "$(Get-Date) RG number: $($list.Count)"; Start-Sleep -Seconds 60}
             azPSVersion: "latest"
```

**Comment by matsest** [comment]
FWIW I was able to test this here with 1hr time between operations, using a managed identity.

Link to action: https://github.com/matsest/az-oidc-managed-identity-demo/actions/runs/3896459977/jobs/6653023616

Workflow definition: https://github.com/matsest/az-oidc-managed-identity-demo/blob/37467e85623d9472885a52293e6bf9ca20ce994e/.github/workflows/login.yaml

![image](https://user-images.githubusercontent.com/10927304/212723175-0686c4c3-c5d5-4c31-89e1-1934b94a8292.png)

**Comment by sandrochristiaan** [comment]
@dingmeng-xue Thanks for your time & attention on this topic.

What @krowlandson is describing is exactly the use case.
I have multiple actions configured after the azure login action.
After some time the token indeed will expire.

![image](https://user-images.githubusercontent.com/22005405/212730118-7af254f3-acfc-4187-bc52-34de57c09f53.png)

![image](https://user-images.githubusercontent.com/22005405/212730801-5b0936f4-cb4e-4a70-a768-36f494cafdff.png)

Is this by design?

I guess the token lifetime expiration issue has been fixed. But there is no functionality in place to do auto-refresh of the token.

Can you confirm the managed identities configuration as tested by @matsest is the only way to go when using long running workflows in combination with OIDC auth?

**Comment by matsest** [comment]
> Can you confirm the managed identities configuration as tested by @matsest is the only way to go when using long running workflows in combination with OIDC auth?

@sandrochristiaan  From your screenshots it looks like you're using Az Module 7.5.0? Please try to upgrade to 9.2 or newer (Az.Accounts 2.10.4 or newer).

**Comment by sandrochristiaan** [comment]
> > Can you confirm the managed identities configuration as tested by @matsest is the only way to go when using long running workflows in combination with OIDC auth?
> 
> @sandrochristiaan From your screenshots it looks like you're using Az Module 7.5.0? Please try to upgrade to 9.2 or newer (Az.Accounts 2.10.4 or newer).

I use GitHub hosted runners, should I update manually??
Guess there is no other way :/

**Comment by matsest** [comment]
@sandrochristiaan

You should identify which runner image you run on to verify which Az version is bundled (example for `ubuntu-22.04`: https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md#powershell-modules). Also make sure you're [specifying either 'latest' or '9.2.0' as a version](https://github.com/Azure/powershell#available-versions-of-az-module-on-runner) in your azure/powershell action if that version is supported. If not, you should update/install it inline in your workflow.

It would also be helpful to share workflow definition files (either as link / gist / inline). See a link to the workflow definition I used [above](https://github.com/Azure/login/issues/180#issuecomment-1384270328) (I'm not using the azure/powershell action, just the login action).

**Comment by BartDecker** [comment]
Hi @matsest . I'm took this issue over within the company so going to have a look at it in the upcoming days. Had a little bit of a hand over but still need to read up on the issue.

**Comment by BartDecker** [comment]
@matsest I can also confirm that the issue is still there. The Az.Accounts used for the login action = 2.11.1

**Comment by isra-fel** [comment]
I tried with @matsest 's workflow with OIDC+SP. The tokens had 24h lifespan.
![image](https://user-images.githubusercontent.com/11371776/217783459-fbf2ebe4-4d90-4d8b-8a42-d4d5554c331d.png)
Workflow: https://github.com/isra-fel/az-oidc-managed-identity-demo/blob/dev-sp/.github/workflows/login-sp-access-tokens.yaml

AzVersion was "2.11.1" in my case, but the fix should be there since 2.10.4.

P.S. They did mention that [This expiration time is further configurable in Azure](https://github.com/Azure/login#:~:text=This%20expiration%20time%20is%20further%20configurable%20in%20Azure). Could that be the reason of the error?

**Comment by BartDecker** [comment]
> I tried with @matsest 's workflow with OIDC+SP. The tokens had 24h lifespan. ![image](https://user-images.githubusercontent.com/11371776/217783459-fbf2ebe4-4d90-4d8b-8a42-d4d5554c331d.png) Workflow: https://github.com/isra-fel/az-oidc-managed-identity-demo/blob/dev-sp/.github/workflows/login-sp-access-tokens.yaml
> 
> AzVersion was "2.11.1" in my case, but the fix should be there since 2.10.4.
> 
> P.S. They did mention that [This expiration time is further configurable in Azure](https://github.com/Azure/login#:~:text=This%20expiration%20time%20is%20further%20configurable%20in%20Azure). Could that be the reason of the error?

Hi @isra-fel is this now a test using a SP or a managed identity? because the screenshots shows oidc-managed-identity-demo etc

**Comment by BartDecker** [comment]
bump to keep topic active

**Comment by MiArth** [comment]
this is in fact a show stopper for many business use cases - any info about the state dear azure team?

**Comment by BartDecker** [comment]
> AzVersion was "2.11.1" in my case, but the fix should be there since 2.10.4.
> 
> P.S. They did mention that [This expiration time is further configurable in Azure](https://github.com/Azure/login#:~:text=This%20expiration%20time%20is%20further%20configurable%20in%20Azure). Could that be the reason of the error?

This is not the way to test the issue at hand.. The proper way to test is to just have the login action (OIDC) and have a a  job which just run a little but longer.

**When I insert your test in between the below workflow the get-accesstoken has a validity of:**

![image](https://user-images.githubusercontent.com/107387835/225540281-e78285f0-d0c6-4ebd-bef6-4881fc384773.png)

**But as you can see the workflow/job fails after several minutes anyways with:**

: Client assertion is not
     | within its valid time range. Current time: 2023-03-16T06:46:21.3974707Z,
     | assertion valid from 2023-03-16T06:28:41.0000000Z, expiry time of
     | assertion 2023-03-16T06:33:41.0000000Z 

![image](https://user-images.githubusercontent.com/107387835/225539425-332e3387-bb2f-4c27-84ee-912d25b51692.png)

**Comment by BartDecker** [comment]
@dcaro Hello Damien, is there awareness about this bug or is it being worked on? Seems to be quite a community effort to keep the topic alive but there seems to be no 'official' status updates on the whole thing. 

It's maybe better to label it as "by design" and update the respective documentation if nothing is done about it.

**Comment by dcaro** [comment]
@BartDecker we are aware of the issue and working on it. 
Do you have a scenario that you can share with us that would allow us to reproduce the issue all the time.

**Comment by sabbour** [comment]
I'm facing this issue as well. Why is this duration not configurable?

**Comment by matsest** [comment]
@dcaro > we are aware of the issue and working on it. Do you have a scenario that you can share with us that would allow us to reproduce the issue all the time.

Something like this (with configurable times to run in parallel have made it possible to reproduce it for me:

https://github.com/matsest/az-oidc-managed-identity-demo/blob/4da25b3055dd4254f247fc8f8dc308bc9af07bc6/.github/workflows/login.yaml

**Comment by sabbour** [comment]
Similar issue in another tool that was fixed by refetching the token instead of storing it: https://github.com/Azure/azure-dev/pull/1963

**Comment by BartDecker** [comment]
> Similar issue in another tool that was fixed by refetching the token instead of storing it: [Azure/azure-dev#1963](https://github.com/Azure/azure-dev/pull/1963)

Hi @sabbour Thanks for linking. Reading the linked topics I don't understand if I need to adjust something to test this or if there was a PR fixing OIDC token handling?

**Comment by YanaXu** [comment]
Hi @matsest and all,

I can reproduce the issue as matsest's [comment](https://github.com/Azure/login/issues/180#issuecomment-1384270328).
I use OIDC with a service principal.
My test workflow is :
```yaml
name: Run Azure Login Dev - SP
on: [push]

permissions:
  id-token: write
  contents: read
jobs:
  demo:
    runs-on: ubuntu-latest
    continue-on-error: true
    strategy:
      matrix:
        sleep: [2401, 3001, 3301, 3481, 3601] # 40m, 50m, 55m, 58m, 60m
    steps:
      - name: "Azure PowerShell login"
        uses: azure/login@v1
        with:
          client-id: ${{ secrets.AZURE_CLIENT_ID }}
          tenant-id: ${{ secrets.AZURE_TENANT_ID }}
          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
          enable-AzPSSession: true

      - name: "Install Az modules"
        run: |
            Set-PSRepository -Name 'PSGallery' -InstallationPolicy Trusted
            Install-Module -Name Az.Resources
            Get-Module -ListAvailable -Name "Az.*" | Select Name, Version
        shell: pwsh
 
      - name: "Test length of token"
        run: |
            $errorActionPreference='Continue'

            Get-AzResourceGroup | Select -First 1

            $now = Get-Date -AsUTC
            Write-Host "Now: $now"
            $token = Get-AzAccessToken
            Write-Host "Expires: $($token.ExpiresOn.DateTime)"
            
            Write-Host "Sleeping for ${{ matrix.sleep }} seconds.."
            Start-Sleep -Seconds ${{ matrix.sleep }}
            Write-Host "Sleep stop!"

            $now = Get-Date -AsUTC
            Write-Host "Now: $now"

            Get-AzResourceGroup | Select -First 1

            $now = Get-Date -AsUTC
            Write-Host "Now: $now"
            $token = Get-AzAccessToken
            Write-Host "Expires: $($token.ExpiresOn.DateTime)"
        shell: pwsh
```
**The test result is:**

- When sleep time is <=3001(50min), the workflow is working fine.
- When sleep time is >=3301(55min), the workflow failes.

The error message for sleep time 3301 is:
```PowerShell
ResourceGroupName : ****************
Location          : ****************
ProvisioningState : Succeeded
Tags              : 
ResourceId        : ****************

Now: 04/26/2023 07:37:26
Expires: 04/26/2023 08:37:12
Sleeping for 3301 seconds..
Sleep stop!
Now: 04/26/2023 08:32:27
Get-AzResourceGroup: /home/runner/work/_temp/fdbc8064-3d0f-4a74-8594-0214a0846c74.ps1:18
Line |
  18 |  Get-AzResourceGroup | Select -First 1
     |  ~~~~~~~~~~~~~~~~~~~
     | Your Azure credentials have not been set up or have expired, please run
     | Connect-AzAccount to set up your Azure credentials. A configuration
     | issue is preventing authentication - check the error message from the
     | server for details. You can modify the configuration in the application
     | registration portal. See https://aka.ms/msal-net-invalid-client for
     | details.  Original exception: AADSTS700024: Client assertion is not
     | within its valid time range. Current time: 2023-04-26T08:32:27.6947912Z,
     | assertion valid from 2023-04-26T07:36:40.0000000Z, expiry time of
     | assertion 2023-04-26T07:41:40.0000000Z. Review the documentation at
     | https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials . Trace ID: 3a3b4a07-1445-43d6-9265-6eaa22404e00 Correlation ID: 427b169e-c228-4a29-8219-6503a46e330a Timestamp: 2023-04-26 08:32:27Z

Now: 04/26/2023 08:32:27
Get-AzAccessToken: /home/runner/work/_temp/fdbc8064-3d0f-4a74-8594-0214a0846c74.ps1:22
Line |
  22 |  $token = Get-AzAccessToken
     |           ~~~~~~~~~~~~~~~~~
     | A configuration issue is preventing authentication - check the error
     | message from the server for details. You can modify the configuration in
     | the application registration portal. See
     | https://aka.ms/msal-net-invalid-client for details.  Original exception:
     | AADSTS700024: Client assertion is not within its valid time range.
     | Current time: 2023-04-26T08:32:27.9421540Z, assertion valid from
     | 2023-04-26T07:36:40.0000000Z, expiry time of assertion
     | 2023-04-26T07:41:40.0000000Z. Review the documentation at
     | https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials . Trace ID: 3a3b4a07-1445-43d6-9265-6eaa28404e00 Correlation ID: 9004a59d-d789-4d24-b0a9-05926de3d3ef Timestamp: 2023-04-26 08:32:27Z

Expires: 

```
Before sleep, the access token shows it'll expire at "04/26/2023 08:37:12"(1 hour later). But after sleep, the time is "04/26/2023 08:32:27" (still in the valid hour), but `Get-AzResourceGroup` failes.

**The issue I understand here is:**
When login with OIDC with a service principal, the token will be expired in 55 min(less than 1 hour ).
If you have other issues, please comment and @me.

**I also need to confirm your requirement. Which option is what you want?**

- The token for OIDC with a service principal should be expired after 1 hour.
- The token for OIDC with a service principal should be expired after 24 hours, same as with managed identity.

**Comment by BartDecker** [comment]
> ```yaml
> $token = Get-AzAccessToken
> ```

Hi @YanaXu  

Thanks for testing but the test is not valid for the problem. Probably $token = Get-AzAccessToken gets a new token after than the initial OICD token. 

See my earlier comments here: https://github.com/Azure/login/issues/180#issuecomment-1471425657

**Comment by YanaXu** [comment]
Hi @BartDecker ,

JFY, I also test with the action part as
```yaml
      - name: "Test length of token"
        run: |
            Get-AzResourceGroup | Select -First 1
            Write-Host "Sleeping for ${{ matrix.sleep }} seconds.."
            Start-Sleep -Seconds ${{ matrix.sleep }}
            Write-Host "Sleep stop!"
            Get-AzResourceGroup | Select -First 1
        shell: pwsh
```

No `Get-AzAccessToken` in the workflow.
And the result is the same.
![image](https://user-images.githubusercontent.com/12843157/234534615-ab12f635-828d-4da3-838e-5ce5a258fc86.png)

**Comment by YanaXu** [comment]
> > ```yaml
> > $token = Get-AzAccessToken
> > ```
> 
> Hi @YanaXu Yan Xu FTE
> 
> Thanks for testing but the test is not valid for the problem. Probably $token = Get-AzAccessToken gets a new token after getting the initial OICD token.
> 
> We can have a screen sharing session if yo want. Would be nice to get to the bottom of this as there is some confusion in this discussion every now and than.
> 
> See my earlier comments here: [#180 (comment)](https://github.com/Azure/login/issues/180#issuecomment-1471425657)

@BartDecker , It's better to share your workflow file and the related workflow output. If it's not OK to share, please ping me on [teams](https://www.microsoft.com/en-us/microsoft-teams/log-in), and we can schedule a meeting. You can find my teams account (mail) on GitHub profile. Thanks.

**Comment by BartDecker** [comment]
@YanaXu many thanks for the opportunity to reach out to you on teams.

I have retested and it indeed works as expected. 

**Some notes for others who run into the same problem:**

* It's not only the az version that belongs to the login step of your job which counts
* we had several steps that can't use the latest azpwsh version but use azpwsh 7.5.0 These jobs will fail and actually need a re-login before they start. 
* Our 7.5.0 jobs take less than 5 mins which make a re-login work for these steps. It could be that the jobs would fail if it takes longer than 5 mins. 

Tested with a workflow consisting out of multiple steps which use :  azPSVersion: "latest" for the deployment steps and "AzVersion": "2.12.1" for the login step.

**Comment by matsest** [comment]
> @BartDecker: indeed works as expected.

Could you elaborate? It would also be helpful if you can share a full workflow definition, preferably with a link to a run as well.

> It could be that the jobs would fail if it takes longer than 5 mins.

To me this still seems to be the main issue here.

**Comment by YanaXu** [comment]
As I can see, there are 2 issues:
- The token will be expired in 5 min.
  - This is the original issue, and it's fixed in Azure PowerShell v9.2 (released on 12/6/2022).
  - It's also the issue [BartDecker](https://github.com/BartDecker) mentioned. 
- The token will be expired after 50 min.
  - It's the issue as your [comment  ](https://github.com/Azure/login/issues/180#issuecomment-1384270328). And I reproduce it in  [comment](https://github.com/Azure/login/issues/180#issuecomment-1523079675).

:exclamation::exclamation::exclamation:**Here are the notes for all:
:exclamation::exclamation::exclamation:Please upgrade your Azure PowerShell Modules version to the latest! NOT only the Azure Login Action but ALL actions (if your action runs any Azure PowerShell Module), e.g., Azure PowerShell Action, PowerShell scripts to run Azure PowerShell, ...**

@matsest , please kindly check my above question in this [comment](https://github.com/Azure/login/issues/180#issuecomment-1523079675).

**Comment by BartDecker** [comment]
> Has anyone found a workaround for the tokens that expire after ~1h? I guess you could detect that case and run the login action again?

Indeed just a re-login. We look at the average runtime of a workflow and put a re-login at the appropriate point in the workflow. Would be nice if the token time is configurable as part of the login action.

**Comment by dionhaefner** [comment]
I ended up just using a shell script `login.sh` for OIDC login (instead of the login action):

```bash
#!/bin/bash
set -e

# get JWT from GitHub's OIDC provider
# see https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#updating-your-actions-for-oidc
jwt_token=$(
    curl \
        -H "Authorization: bearer $ACTIONS_ID_TOKEN_REQUEST_TOKEN" \
        "$ACTIONS_ID_TOKEN_REQUEST_URL&audience=api://AzureADTokenExchange" \
        --silent \
    | jq -r ".value"
)

# perform OIDC token exchange
az login \
    --service-principal -u $AZURE_CLIENT_ID \
    --tenant $AZURE_TENANT_ID \
    --federated-token $jwt_token \
    -o none

az account set \
    --subscription $AZURE_SUBSCRIPTION_ID \
    -o none
```

Then you can keep track of time in long-running scripts and renew the token every 50 minutes or so.

**Comment by surajssd** [comment]
Hit this issue on Ubuntu based workflow. We login upfront, then try to do an image build (which runs for 10 mins), meanwhile the token is expired and Hashicorp packer fails because the token has expired.

<img width="1025" alt="image" src="https://github.com/Azure/login/assets/5815795/aa9c9524-d46b-4d95-8e8b-83fa1d1767ec">

Here is the log from that failed step:

```bash
...
peer-pod-ubuntu.azure-arm.ubuntu: output will be in this color.

==> peer-pod-ubuntu.azure-arm.ubuntu: Running builder ...
==> peer-pod-ubuntu.azure-arm.ubuntu: Getting tokens using Azure CLI
==> peer-pod-ubuntu.azure-arm.ubuntu: unable to get token from azure cli: Invoking Azure CLI failed with the following error: ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2023-05-17T01:01:18.0483732Z, assertion valid from 2023-05-17T00:45:33.0000000Z, expiry time of assertion 2023-05-17T00:50:33.0000000Z. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials .
==> peer-pod-ubuntu.azure-arm.ubuntu: Trace ID: 2b484d1e-c5db-483f-9502-c274b6715c00
==> peer-pod-ubuntu.azure-arm.ubuntu: Correlation ID: 97a5e8bf-8386-4f13-8db0-8cc9a18f4fad
==> peer-pod-ubuntu.azure-arm.ubuntu: Timestamp: 2023-05-17 01:01:18Z
==> peer-pod-ubuntu.azure-arm.ubuntu: Interactive authentication is needed. Please run:
==> peer-pod-ubuntu.azure-arm.ubuntu: az login
==> peer-pod-ubuntu.azure-arm.ubuntu:
Build 'peer-pod-ubuntu.azure-arm.ubuntu' errored after 1 second 393 milliseconds: Invoking Azure CLI failed with the following error: ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2023-05-17T01:01:18.0483732Z, assertion valid from 2023-05-17T00:45:33.0000000Z, expiry time of assertion 2023-05-17T00:50:33.0000000Z. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials .
Trace ID: 2b484d1e-c5db-483f-9502-c274b6715c00
Correlation ID: 97a5e8bf-8386-4f13-8db0-8cc9a18f4fad
Timestamp: 2023-05-17 01:01:18Z
Interactive authentication is needed. Please run:
az login


==> Wait completed after 1 second 393 milliseconds

==> Some builds didn't complete successfully and had errors:
--> peer-pod-ubuntu.azure-arm.ubuntu: Invoking Azure CLI failed with the following error: ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2023-05-17T01:01:18.0483732Z, assertion valid from 2023-05-17T00:45:33.0000000Z, expiry time of assertion 2023-05-17T00:50:33.0000000Z. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials .
Trace ID: 2b484d1e-c5db-483f-9502-c274b6715c00
Correlation ID: 97a5e8bf-8386-4f13-8db0-8cc9a18f4fad
Timestamp: 2023-05-17 01:01:18Z
Interactive authentication is needed. Please run:
az login


==> Builds finished but no artifacts were created.
make: *** [Makefile:25: podvm-c491ecd-amd64] Error 1
Error: Process completed with exit code 2.

```

The [docs](https://github.com/Azure/login#:~:text=This%20expiration%20time%20is%20further%20configurable%20in%20Azure) say that the expiry time is configurable. But they don't say how.

> This expiration time is further configurable in Azure

**Comment by YanaXu** [comment]
Hi @surajssd , have you checked my [comment](https://github.com/Azure/login/issues/180#issuecomment-1524995605)? Have you upgraded all Azure PowerShell versions, including the manually installed one or Azure PowerShell Action?

**Comment by surajssd** [comment]
@YanaXu not using Powershell. Only Ubuntu/Linux based actions.

**Comment by YanaXu** [comment]
@surajssd , in this issue, we're talking about the Azure PowerShell login issue with OIDC with the service principle. If you do not use Azure PowerShell at all, we are meeting another issue. Let's see what's wrong in your case.
Could you share the workflow with me? You can mask the sensitive information, but let me know which actions and what general parameters you are using.

**Comment by mestrada-redstage** [comment]
I have this error when when the build take more that 10 min
![image](https://github.com/Azure/login/assets/95379510/a274aa2c-9901-46f1-aa66-9ffe9d3baff5)
runner: ubuntu-latest
and use manage identity for login into my AKS cluster

**Comment by YanaXu** [comment]
@mestrada-redstage, can you share the workflow with us?

**Comment by mestrada-redstage** [comment]
@YanaXu This are the steps
steps:
        
      - name: Checkout
        uses: actions/checkout@v3

      - name: Set env branch short name
        run: echo "BRANCH_NAME=${TRIGGER_GITHUB_REF#refs/heads/}" >> $GITHUB_ENV

      # setup Docker buld action
      - name: Set up Docker Buildx
        id: buildx
        uses: docker/setup-buildx-action@v1

      - name: Login to DockerHub
        uses: docker/login-action@v2
        with:
          registry: ${{ secrets.ACR_URL }}
          username: ${{ secrets.ACR_LOGIN }}
          password: ${{ secrets.ACR_KEY }} 

      - name: Setup anvironemnt name
        run: |
         echo "ENV_STR=${BRANCH_NAME/development/dev}" >> $GITHUB_ENV
         cat $GITHUB_ENV

      - uses: azure/login@v1
        with:
          client-id: ${{ secrets.AZURE_CLIENT_ID }}
          tenant-id: ${{ secrets.AZURE_TENANT_ID }}
          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
          
      - uses: azure/use-kubelogin@v1
        with:
          kubelogin-version: 'v0.0.24'
          
      - uses: azure/setup-kubectl@v3
       
      - uses: azure/aks-set-context@v3
        with:
          resource-group: #########
          cluster-name: #######

      - name: build-deploy
        shell: bash
        run: |
          cd ${{ matrix.PROJECT_PATH }}
          chmod +x ../build.sh
          ACR_URL=${{ secrets.ACR_URL }} ACR_KEY=${{ secrets.ACR_KEY }} ACR_LOGIN=${{ secrets.ACR_LOGIN }} ../build.sh ${{ env.PROJECT_NAME }} ${{ env.TRIGGER_GITHUB_SHA }} ${{ github.event.inputs.destin }} ${{ env.PROJECT_NAME }} ${{ matrix.PROJECT_CONTEXT }}

**Comment by YanaXu** [comment]
@mestrada-redstage ,

In this issue, we talked about Azure PowerShell login problem with OIDC. 
But your workflow seems using Azure Cli.
I think you were talking about this step which failed after 10 min.
```yaml
  - name: build-deploy
    shell: bash
    run: |
      cd ${{ matrix.PROJECT_PATH }}
      chmod +x ../build.sh
      ACR_URL=${{ secrets.ACR_URL }} ACR_KEY=${{ secrets.ACR_KEY }} ACR_LOGIN=${{ secrets.ACR_LOGIN }} ../build.sh ${{ env.PROJECT_NAME }} ${{ env.TRIGGER_GITHUB_SHA }} ${{ github.event.inputs.destin }} ${{ env.PROJECT_NAME }} ${{ matrix.PROJECT_CONTEXT }}
```

Could you re-run this workflow with **debug** mode and copy the workflow's output to us? 
And if it's not sensitive, please provide the content of "build.sh" too.
Thanks.

**Comment by jbelien** [comment]
Also hitting the same issue.
Here is the log (with debug):

```log
##[debug]Evaluating condition for step: 'Get K8s context'
##[debug]Evaluating: success()
##[debug]Evaluating success:
##[debug]=> true
##[debug]Result: true
##[debug]Starting: Get K8s context
##[debug]Loading inputs
##[debug]Evaluating: vars.AZURE_RESOURCE_GROUP
##[debug]Evaluating Index:
##[debug]..Evaluating vars:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'AZURE_RESOURCE_GROUP'
##[debug]=> 'rg-***'
##[debug]Result: 'rg-***'
##[debug]Evaluating: vars.AKS_CLUSTER_NAME
##[debug]Evaluating Index:
##[debug]..Evaluating vars:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'AKS_CLUSTER_NAME'
##[debug]=> 'aks-***'
##[debug]Result: 'aks-***'
##[debug]Loading env
Run azure/aks-set-context@v3
  with:
    resource-group: rg-***
    cluster-name: aks-***
    admin: false
    use-kubelogin: true
  env:
    CONTAINER_NAME: ***
    AZURE_HTTP_USER_AGENT: 
    AZUREPS_HOST_ENVIRONMENT: 
##[debug]Writing kubeconfig to /home/runner-admin/actions-runner/_work/_temp/kubeconfig_1691671866005
/usr/bin/az aks get-credentials --resource-group rg-*** --name aks-*** -f /home/runner-admin/actions-runner/_work/_temp/kubeconfig_1691671866005
ERROR: AADSTS700024: Client assertion is not within its valid time range. Current time: 2023-08-10T12:51:06.6841503Z, assertion valid from 2023-08-09T15:06:56.0000000Z, expiry time of assertion 2023-08-09T15:11:56.0000000Z. Review the documentation at https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials .
Trace ID: 070e96fc-76e1-46a8-88f7-055a0d489900
Correlation ID: 2ce46b3e-5dcf-42ed-a497-33231a9dc308
Timestamp: 2023-08-10 12:51:06Z
Interactive authentication is needed. Please run:
az login
Error: Error: The process '/usr/bin/az' failed with exit code 1
##[debug]Node Action run completed with exit code 1
##[debug]AZURE_HTTP_USER_AGENT='GitHubActions/azure/aks-set-context(6ff0b0b4611939515451a12aa8fec15e3dd86d30d7e8c0569dd2ded2113f8931; 5809746149)'
##[debug]AZUREPS_HOST_ENVIRONMENT='GitHubActions/azure/aks-set-context(6ff0b0b4611939515451a12aa8fec15e3dd86d30d7e8c0569dd2ded2113f8931; 5809746149)'
##[debug]AZURE_HTTP_USER_AGENT=''
##[debug]AZUREPS_HOST_ENVIRONMENT=''
##[debug]Finishing: Get K8s context
```

**Comment by hansenms** [comment]
I just wanted to chime in here, we are experiencing the same problem that the client assertion expires after 5 minutes and `az aks get-credentials` is failing in the same way as described above. Other `az` commands work, so it is possible that the token is not refreshed for that command?

We use a script (along the lines of https://github.com/Azure/login/issues/180#issuecomment-1545459942 by @dionhaefner) right in front of calls to `az aks get-credentials` in our pipelines to work around this.

**Comment by mcolussi-bigw** [comment]
Issue is occurring for us as well. We use null_resource terraform resources to run custom scripts that perform commands via az cli, the automated tests are failing because the token expires after 5 minutes, before the scripts execute.

Our current only work-around is to ensure the tests that run AZ cli commands run first before the token expires. 5 minutes is a very short time range to do anything useful.

**Comment by JamesDawson** [comment]
Are there any plans to support idTokens with a lifetime longer than 10 minutes?  This constraint has sadly forced us back to using SP authentication with client secrets.

We have a scripted deployment process (i.e. runs as a single GitHub Actions step using `azure/powershell@v1`) that runs for longer than 10 minutes.  It includes functionality towards the end that requires an access token from another service (i.e. a data plane operation), which fails as the idToken has expired.

As a tangible example of this scenario, consider using the `Az.Synapse` PowerShell module to make configuration changes to a Synapse Workspace that has just been provisioned by a preceding ARM deployment (lasting longer than 10 minutes).

**Comment by YanaXu** [comment]
Hi @mestrada-redstage , @jbelien, @hansenms, @mcolussi-bigw , could you open another issue and provide your workflow file and the debug log? Because this issue is about the Azure PowerShell and it's fixed in Azure PowerShell 9.2 as in [comment](https://github.com/Azure/login/issues/180#issuecomment-1524995605). But in your description, it seems there are other issues in Azure CLI or even other actions. Please open a new issue for your case for better tracking.

**Comment by YanaXu** [comment]
> Are there any plans to support idTokens with a lifetime longer than 10 minutes? This constraint has sadly forced us back to using SP authentication with client secrets.
> 
> We have a scripted deployment process (i.e. runs as a single GitHub Actions step using `azure/powershell@v1`) that runs for longer than 10 minutes. It includes functionality towards the end that requires an access token from another service (i.e. a data plane operation), which fails as the idToken has expired.
> 
> As a tangible example of this scenario, consider using the `Az.Synapse` PowerShell module to make configuration changes to a Synapse Workspace that has just been provisioned by a preceding ARM deployment (lasting longer than 10 minutes).
> 
> UPDATE: When testing an [equivalent repro](https://devblogs.microsoft.com/devops/public-preview-of-workload-identity-federation-for-azure-pipelines/#comment-4001) with the recently announced preview of federated credentials for Azure DevOps, we do not see the same issue.

@JamesDawson , do you mean the toke generated by GitHub? I think it's out of the scope of this Action. Can you raise an issue for GitHub Action Core?

**Comment by YanaXu** [comment]
I'll close this issue.

- If you meet an Azure PowerShell token expiration issue, please refer to [comment ](https://github.com/Azure/login/issues/180#issuecomment-1524995605). 
- If you meet other token expiration issues, please open a new issue for it. Provide your workflow file and debug logs for further analysis.

**Comment by jbelien** [comment]
> Hi @mestrada-redstage , @jbelien, @hansenms, @mcolussi-bigw , could you open another issue and provide your workflow file and the debug log? Because this issue is about the Azure PowerShell and it's fixed in Azure PowerShell 9.2 as in [comment](https://github.com/Azure/login/issues/180#issuecomment-1524995605). But in your description, it seems there are other issues in Azure CLI or even other actions. Please open a new issue for your case for better tracking.

Sure, I already opened an issue here: https://github.com/Azure/aks-set-context/issues/101

**Comment by jantorep** [comment]
Any update on this case? This is driving me mad creating Enterprise Applications for our deployments.

New-AzADApplication_CreateExpanded: /home/runner/.local/share/powershell/Modules/Az.Resources/6.12.1/MSGraph.Autorest/custom/New-AzADApplication.ps1:698
Line |
 698 |      $app = Az.MSGraph.internal\New-AzADApplication @PSBoundParameters
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | A configuration issue is preventing authentication - check the error
     | message from the server for details. You can modify the configuration in
     | the application registration portal. See
     | https://aka.ms/msal-net-invalid-client for details.  Original exception:
     | AADSTS[70](https://github.com/HemitHF/AzureIPAM/actions/runs/7299500067/job/19892437562#step:7:71)0024: Client assertion is not within its valid time range.
     | Current time: 2023-12-22T12:03:41.4945531Z, assertion valid from
     | 2023-12-22T11:48:18.0000000Z, expiry time of assertion
     | 2023-12-22T11:53:18.0000000Z. Review the documentation at
     | https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials . Trace ID: e36c2ebe-03f4-427b-b341-9b346cd1bb00 Correlation ID: [71](https://github.com/HemitHF/AzureIPAM/actions/runs/7299500067/job/19892437562#step:7:72)ce361d-10ed-4080-a049-6e15eb7bdee9 Timestamp: 2023-12-22 12:03:41Z

**Comment by YanaXu** [comment]
Hi @krukowskid and @jantorep , it's a closed issue. We may miss comments from a closed issue. Please check [my comment](https://github.com/Azure/login/issues/180#issuecomment-1724916868) for this issue. And If your issue is not solved,  it means it's a new issue. Please open a new issue then, provide your workflow file, debug log for further analysis. Thanks.

**Comment by justinmchase** [comment]
Utmost respect for the maddest lad @dionhaefner for this comment https://github.com/Azure/login/issues/180#issuecomment-1545459942

Which allowed me to work around this issue in terraform with the following:

```tf
data "azurerm_client_config" "current" {
}

data "external" "login" {
  program = ["bash", "${path.module}/scripts/refresh.sh"]
  query = {
    client_id       = data.azurerm_client_config.current.client_id
    tenant_id       = data.azurerm_client_config.current.tenant_id
    subscription_id = data.azurerm_client_config.current.subscription_id
  }
}
```
```tf
#!/bin/bash
set -e

eval "$(jq -r '@sh "AZURE_CLIENT_ID=\(.client_id) AZURE_TENANT_ID=\(.tenant_id) AZURE_SUBSCRIPTION_ID=\(.subscription_id)"')"

if [[ "$ACTIONS_ID_TOKEN_REQUEST_TOKEN" == "" ]] || [[ "$ACTIONS_ID_TOKEN_REQUEST_URL" == "" ]]
then
  # This is ok, we're probably running locally
  jq -n '{}'
  exit 0
fi

# get JWT from GitHub's OIDC provider
# see https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#updating-your-actions-for-oidc
jwt_token=$(
  curl "$ACTIONS_ID_TOKEN_REQUEST_URL&audience=api://AzureADTokenExchange" \
    -H "Authorization: bearer $ACTIONS_ID_TOKEN_REQUEST_TOKEN" \
    --silent \
  | jq -r ".value"
)

# perform OIDC token exchange
az login \
  --service-principal -u $AZURE_CLIENT_ID \
  --tenant $AZURE_TENANT_ID \
  --federated-token $jwt_token \
  -o none

az account set \
  --subscription $AZURE_SUBSCRIPTION_ID \
  -o none

jq -n '{}'
```

This allows me to refresh the token in the middle of a long terraform apply in a few places where I need to bust out into a `null_resource` or `data external` script and use the az cli.

```tf
module "auth_refresh" {
  source = "../../helpers/azure_auth_refresh"
}

resource "null_resource" "redirect_uris" {
  // ...
}
```
